Jenkinsfile: create junit.xml for integration tests

thaJeztah opened the pull request:
based on top of / depends on:

- [x] https://github.com/moby/moby/pull/39802 integration: simplify parallel run destination

Addresses parts of https://github.com/moby/moby/issues/39675

thaJeztah committed:
integration: use gotestsum to generate junit.xml and go-test-report.json

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

tiborvass commented on hack/make/.integration-test-helpers:
```diff
@@ -67,12 +67,12 @@ run_test_integration() {
 }
 
 run_test_integration_suites() {
+	go get -u github.com/jstemmer/go-junit-report
```
we already have gotestsum installed, let's use that: https://github.com/moby/moby/blob/4e83c90ae8362a8205e3728e8fcb655a713a0d87/hack/dockerfile/install/gotestsum.installer

thaJeztah commented:
One flaky test failed, but otherwise this works; PTAL

I'll squash commits if this looks good

thaJeztah commented on Jenkinsfile:
```diff
@@ -21,6 +21,7 @@ pipeline {
         DOCKER_GRAPHDRIVER  = 'overlay2'
         APT_MIRROR          = 'cdn-fastly.deb.debian.org'
         CHECK_CONFIG_COMMIT = '78405559cfe5987174aa2cb6463b9b2c1b917255'
+        TEST_FILTER         = 'TestBuildMultiStageCopy'
```
oh; need to remove this one as well; was just there for testing

thaJeztah commented:
Starting to look good

<img width="504" alt="Screenshot 2019-08-19 at 22 58 26" src="https://user-images.githubusercontent.com/1804568/63298906-e73e8400-c2d4-11e9-8ddf-2a98d8c4d62f.png">

thaJeztah commented:
<img width="500" alt="Screenshot 2019-08-21 at 19 40 13" src="https://user-images.githubusercontent.com/1804568/63454662-87b6b480-c44b-11e9-9cf2-56fa10b8957c.png">

thaJeztah commented:
@tiborvass @andrewhsu squashed and cleaned up commits; this should be ready to review / ready to go

After this is merged, we'll need to do another round, and implement it for Windows

thaJeztah commented:
This is all green

andrewhsu reviewed (approved):
LGTM

thaJeztah commented:
@tiborvass look here; https://ci.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39724/20/tests

thaJeztah commented:
oh, if you click through in the old UI, you can also see the name spacing based on stage; https://ci.docker.com/public/job/moby/job/PR-39724/20/testReport/tests.integration.api_build_test/BuildTest/

tiborvass commented:
Those new UI is 100% useless and the old UI shows that aside from the Docker-Py tests not being namespaced under a Docker-py namespace, the only tests are unit tests and docker-py tests.

andrewhsu commented:
The new Blue Ocean UI is horrible for browsing 14,876 tests because you have to click on "Show More" to page thru: https://ci.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39724/20/tests

However, when there is a failure, the new Blue Ocean UI will put the test failure at the top of the first page which is helpful.

I find the classic UI easier to navigate to an individual test:
https://ci.docker.com/public/job/moby/job/PR-39724/20/testReport/

For this PR, I think the test names are not showing up because the `testsuite` name in the xml file is blank. e.g. from the `volume-junit-report.xml` file in the `janky-bundles.tar.gz` [artifact](https://ci.docker.com/public/job/moby/job/PR-39724/20/artifact/):
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite tests="5" failures="0" time="0.000000" name="">
    <properties>
      <property name="go.version" value="go1.12.9 linux/amd64"></property>
    </properties>
    <testcase classname="" name="TestVolumesCreateAndList" time="0.000000"></testcase>
    <testcase classname="" name="TestVolumesRemove" time="0.000000"></testcase>
    <testcase classname="" name="TestVolumesInspect" time="0.000000"></testcase>
    <testcase classname="" name="TestVolumesInvalidJSON//volumes/create" time="0.000000"></testcase>
    <testcase classname="" name="TestVolumesInvalidJSON" time="0.000000"></testcase>
  </testsuite>
</testsuites>
```

I think the integration test results are put under the `(root)` section because the xml it has no testsuite name:
https://ci.docker.com/public/job/moby/job/PR-39724/20/testReport/(root)/

For comparison, the master branch tests don't have a `(root)` section:
https://ci.docker.com/public/job/moby/job/master/57/testReport/

thaJeztah commented:
yeah, `gotestsum` would need something similar as the `junit_suite_name` that pytest has http://doc.pytest.org/en/latest/usage.html

andrewhsu commented:
Looks like changing the testsuite name was discussed before: https://github.com/gotestyourself/gotestsum/issues/50

thaJeztah committed:
integration: test2json: enable timestamps to fix zero-time test durations

Without these options set, test2json does not include a `Time`
field in the generated JSON;

    {"Action":"run","Test":"TestCgroupNamespacesBuild"}
    {"Action":"output","Test":"TestCgroupNamespacesBuild","Output":"=== RUN   TestCgroupNamespacesBuild\n"}
    {"Action":"output","Test":"TestCgroupNamespacesBuild","Output":"--- PASS: TestCgroupNamespacesBuild (1.70s)\n"}
    ...
    {"Action":"pass","Test":"TestCgroupNamespacesBuild"}

As a result, `gotestsum` was not able to calculate test-duration, and
reported `time="0.000000"` for all tests;

    <testcase classname="amd64.integration.build" name="TestCgroupNamespacesBuild" time="0.000000"></testcase>

With this patch applied:

    {"Time":"2019-08-23T22:42:41.644361357Z","Action":"run","Package":"amd64.integration.build","Test":"TestCgroupNamespacesBuild"}
    {"Time":"2019-08-23T22:42:41.644367647Z","Action":"output","Package":"amd64.integration.build","Test":"TestCgroupNamespacesBuild","Output":"=== RUN   TestCgroupNamespacesBuild\n"}
    {"Time":"2019-08-23T22:42:44.926933252Z","Action":"output","Package":"amd64.integration.build","Test":"TestCgroupNamespacesBuild","Output":"--- PASS: TestCgroupNamespacesBuild (3.28s)\n"}
    ...
    {"Time":"2019-08-23T22:42:44.927003836Z","Action":"pass","Package":"amd64.integration.build","Test":"TestCgroupNamespacesBuild","Elapsed":3.28}

Which now correctly reports the test's duration:

    <testcase classname="amd64.integration.build" name="TestCgroupNamespacesBuild" time="3.280000"></testcase>

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration: improve package- and filename for junit.xml

Generate more unique names, based on architecture and test-suite name.

Clean up the path to this integration test to create a useful package name.
"$dir" can be either absolute (/go/src/github.com/docker/docker/integration/foo)
or relative (./integration/foo). To account for both, first we strip the
absolute path, then any leading periods and slashes.

For the package-name, we use periods as separator instead of slashes, to be more
in-line with Java package names (which is what junit.xml was originally designed
for).

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Ok, I pushed some changes to;

- fix the reports not having a duration for tests 😅 
- add a package name, so that the tests show up properly in the old Jenkins UI, and are namespaced/grouped (apparently `test2json` has flags to set this 👍)

We can still discuss the best format for the name spacing, but this is what it looks like now;

<img width="1402" alt="Screenshot 2019-08-23 at 17 40 24" src="https://user-images.githubusercontent.com/1804568/63605081-857c6380-c5cd-11e9-8610-b8c48ee7e755.png">

There's some weirdness in showing the individual tests though; I suspect that Jenkins re-scans _all_ junit.xml files for a stage for each file that is uploaded. As a result each test is indexed 4 times (or more);

<img width="827" alt="Screenshot 2019-08-23 at 17 41 14" src="https://user-images.githubusercontent.com/1804568/63605175-b3fa3e80-c5cd-11e9-81f2-e3f03fd5dabd.png">

The total numbers may still be correct (will have to check), but that indexing looks odd

thaJeztah commented:
My suspicion confirmed; we upload 6 files, so I think that's a bug in Jenkins where each upload triggers indexing again (for a stage);

[amd64-integration-build-junit-report.xml.txt](https://github.com/moby/moby/files/3535234/amd64-integration-build-junit-report.xml.txt)
[amd64-integration-plugin-graphdriver-junit-report.xml.txt](https://github.com/moby/moby/files/3535235/amd64-integration-plugin-graphdriver-junit-report.xml.txt)
[amd64-integration-plugin-logging-junit-report.xml.txt](https://github.com/moby/moby/files/3535236/amd64-integration-plugin-logging-junit-report.xml.txt)
[amd64-integration-plugin-volumes-junit-report.xml.txt](https://github.com/moby/moby/files/3535237/amd64-integration-plugin-volumes-junit-report.xml.txt)
[amd64-integration-service-junit-report.xml.txt](https://github.com/moby/moby/files/3535238/amd64-integration-service-junit-report.xml.txt)
[amd64-integration-session-junit-report.xml.txt](https://github.com/moby/moby/files/3535239/amd64-integration-session-junit-report.xml.txt)

thaJeztah committed:
Jenkinsfile: add TESTDEBUG

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Hm; looks like there's still an issue (at least with the uploaded bundles); with debug enabled, I see that these tests are run. Which means that we should have in total **18 go-test-report.json** files, and **18 junit-report.xml** files.


```
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-build-go-test-report.json              --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-build-junit-report.xml              --raw-command -- go tool test2json -p amd64.integration.build              -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-config-go-test-report.json             --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-config-junit-report.xml             --raw-command -- go tool test2json -p amd64.integration.config             -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-container-go-test-report.json          --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-container-junit-report.xml          --raw-command -- go tool test2json -p amd64.integration.container          -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-image-go-test-report.json              --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-image-junit-report.xml              --raw-command -- go tool test2json -p amd64.integration.image              -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-network-go-test-report.json            --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-network-junit-report.xml            --raw-command -- go tool test2json -p amd64.integration.network            -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-network-ipvlan-go-test-report.json     --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-network-ipvlan-junit-report.xml     --raw-command -- go tool test2json -p amd64.integration.network.ipvlan     -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-network-macvlan-go-test-report.json    --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-network-macvlan-junit-report.xml    --raw-command -- go tool test2json -p amd64.integration.network.macvlan    -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-go-test-report.json             --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-junit-report.xml             --raw-command -- go tool test2json -p amd64.integration.plugin             -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-authz-go-test-report.json       --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-authz-junit-report.xml       --raw-command -- go tool test2json -p amd64.integration.plugin.authz       -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-common-go-test-report.json      --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-common-junit-report.xml      --raw-command -- go tool test2json -p amd64.integration.plugin.common      -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-graphdriver-go-test-report.json --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-graphdriver-junit-report.xml --raw-command -- go tool test2json -p amd64.integration.plugin.graphdriver -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-logging-go-test-report.json     --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-logging-junit-report.xml     --raw-command -- go tool test2json -p amd64.integration.plugin.logging     -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-volumes-go-test-report.json     --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-plugin-volumes-junit-report.xml     --raw-command -- go tool test2json -p amd64.integration.plugin.volumes     -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-secret-go-test-report.json             --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-secret-junit-report.xml             --raw-command -- go tool test2json -p amd64.integration.secret             -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-service-go-test-report.json            --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-service-junit-report.xml            --raw-command -- go tool test2json -p amd64.integration.service            -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-session-go-test-report.json            --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-session-junit-report.xml            --raw-command -- go tool test2json -p amd64.integration.session            -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-system-go-test-report.json             --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-system-junit-report.xml             --raw-command -- go tool test2json -p amd64.integration.system             -t ./test.main -test.v -test.timeout=120m
test_env gotestsum --format=standard-verbose --jsonfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-volume-go-test-report.json             --junitfile=/go/src/github.com/docker/docker/bundles/test-integration/amd64-integration-volume-junit-report.xml             --raw-command -- go tool test2json -p amd64.integration.volume             -t ./test.main -test.v -test.timeout=120m
```

However, in the "janky" bundle, I only see; **8 JSON** files, and **6 XML** files, so either the `find` is incorrect, or somewhere we purge files, or files are stored somewhere nested and therefore not found?

```
amd64-integration-config-go-test-report.json
amd64-integration-container-go-test-report.json
amd64-integration-network-ipvlan-go-test-report.json
amd64-integration-network-macvlan-go-test-report.json
amd64-integration-plugin-authz-go-test-report.json
amd64-integration-plugin-go-test-report.json
amd64-integration-plugin-logging-go-test-report.json
amd64-integration-service-go-test-report.json
```

```
amd64-integration-build-junit-report.xml
amd64-integration-plugin-graphdriver-junit-report.xml
amd64-integration-plugin-logging-junit-report.xml
amd64-integration-plugin-volumes-junit-report.xml
amd64-integration-service-junit-report.xml
amd64-integration-session-junit-report.xml
```

thaJeztah commented:
OK, all files _are_ present in the PowerPC and s390x bundles, so I suspect that it's either because the path is different for the Janky run (they use the `1/`, `2/, `3/` subdirectories, or perhaps there's an issue because of the test-flags (it's weird though, because for some tests the _XML_ is there, but the _JSON_ is missing, and vice-versa)

thaJeztah commented:
Trying to run only a single test on Janky (only `integration`, and not the `integration-cli` tests

thaJeztah commented:
Weird; I disabled `unit`, `powerpc` and `z` in the Jenkinsfile, and it's still running those 🤔

thaJeztah commented:
OK, that does solve the problem of only "some" of the files being there; if I only run `integration`, all files are in the bundles.tar.gz;

```
amd64-integration-build-go-test-report.json
amd64-integration-config-go-test-report.json
amd64-integration-container-go-test-report.json
amd64-integration-image-go-test-report.json
amd64-integration-network-go-test-report.json
amd64-integration-network-ipvlan-go-test-report.json
amd64-integration-network-macvlan-go-test-report.json
amd64-integration-plugin-authz-go-test-report.json
amd64-integration-plugin-common-go-test-report.json
amd64-integration-plugin-go-test-report.json
amd64-integration-plugin-graphdriver-go-test-report.json
amd64-integration-plugin-logging-go-test-report.json
amd64-integration-plugin-volumes-go-test-report.json
amd64-integration-secret-go-test-report.json
amd64-integration-service-go-test-report.json
amd64-integration-session-go-test-report.json
amd64-integration-system-go-test-report.json
amd64-integration-volume-go-test-report.json
amd64-integration-build-junit-report.xml
amd64-integration-config-junit-report.xml
amd64-integration-container-junit-report.xml
amd64-integration-image-junit-report.xml
amd64-integration-network-ipvlan-junit-report.xml
amd64-integration-network-junit-report.xml
amd64-integration-network-macvlan-junit-report.xml
amd64-integration-plugin-authz-junit-report.xml
amd64-integration-plugin-common-junit-report.xml
amd64-integration-plugin-graphdriver-junit-report.xml
amd64-integration-plugin-junit-report.xml
amd64-integration-plugin-logging-junit-report.xml
amd64-integration-plugin-volumes-junit-report.xml
amd64-integration-secret-junit-report.xml
amd64-integration-service-junit-report.xml
amd64-integration-session-junit-report.xml
amd64-integration-system-junit-report.xml
amd64-integration-volume-junit-report.xml
```

thaJeztah commented:
Looooool Jenkins now shows each result 36 times 😂 🤷‍♂

andrewhsu commented on Jenkinsfile:
```diff
@@ -302,6 +304,11 @@ pipeline {
                             }
+                            post {
+                                always {
+                                    junit testResults: 'bundles/**/*.xml', allowEmptyResults: true
```
I'm concerned with how the junit plugin processes xml files. It [warns](https://wiki.jenkins.io/display/JENKINS/JUnit+Plugin):
> Be sure not to include any non-report files into this pattern.

And we definitely have xml files that are not report files under the bundles directory:
```
bundles/test-integration/1/TestAuthZPluginV2BadManifestFailsDaemonStart/d26edc53d5db9/root/plugins/20f707871fe12c665ed3685fde0f870ec5d15e1b9d7ee4d154f166ab91944131/rootfs/usr/local/go/doc/codewalk/functions.xml
bundles/test-integration/1/TestAuthZPluginV2BadManifestFailsDaemonStart/d26edc53d5db9/root/plugins/20f707871fe12c665ed3685fde0f870ec5d15e1b9d7ee4d154f166ab91944131/rootfs/usr/local/go/doc/codewalk/markov.xml
bundles/test-integration/1/TestAuthZPluginV2BadManifestFailsDaemonStart/d26edc53d5db9/root/plugins/20f707871fe12c665ed3685fde0f870ec5d15e1b9d7ee4d154f166ab91944131/rootfs/usr/local/go/doc/codewalk/codewalk.xml
bundles/test-integration/1/TestAuthZPluginV2BadManifestFailsDaemonStart/d26edc53d5db9/root/plugins/20f707871fe12c665ed3685fde0f870ec5d15e1b9d7ee4d154f166ab91944131/rootfs/usr/local/go/doc/codewalk/sharemem.xml
```

I suggest chaging the glob to something like `bundles/**/*-junit-report.xml`

thaJeztah replied:
Good catch! Yes, `.xml` files could be present at this point, because container/plugin rootfs's may still be around.

I'll update that matching

thaJeztah commented on Jenkinsfile:
```diff
@@ -299,9 +301,20 @@ pipeline {
+                                # merge artifacts of the three sets
+                                cd "$WORKSPACE/bundles/"
+                                mkdir -p test-integration
+                                mv ./{1..3}/test-integration/* ./test-integration/
```
Hm.. that didn't work;

```
[2019-08-24T12:07:53.797Z] mkdir: cannot create directory ‘test-integration’: Permission denied
[2019-08-24T12:07:53.797Z] mv: target './test-integration/' is not a directory
```

thaJeztah commented:
Splitting the first three commits to a separate PR, as they're not directly related to this change: https://github.com/moby/moby/pull/39790

thaJeztah commented on Jenkinsfile:
```diff
@@ -317,13 +329,22 @@ pipeline {
+                                # merge artifacts of the three sets
+                                cd "$WORKSPACE/bundles/"
+                                mkdir -p test-integration
+                                mv ./{1..3}/test-integration/* ./test-integration/
```
looks like I need bash for this

andrewhsu commented on Jenkinsfile:
```diff
@@ -396,6 +418,11 @@ pipeline {
                             }
+                            post {
+                                always {
+                                    junit testResults: 'bundles/**/*-report.xml', allowEmptyResults: true
```
I believe this glob pattern is what's causing the duplicate results. There is a `latest` symlink that ends up going down a rabbit hole.

See one of the [artifacts](https://ci.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39789/4/artifacts) that are saved with the same glob pattern:
```
bundles/latest/latest/latest/latest/latest/test-integration/amd64-integration-build-junit-report.xml
```

I suggest changing the glob pattern to `bundles/test-integration/*-junit-report.xml` like I did for this test job that just collects tests for `integration`: https://ci.docker.com/public/job/moby/job/PR-39789/6/testReport/

thaJeztah replied:
LOL, wow, nice find! I didn't expect it to be recursing like that (although I've had cases where my IDE wasn't fond of it and got into a loop).

Using `bundles/test-integration/*-junit-report.xml` won't work for the Janky tests, where `1/`, `2/`, `3/` subdirectories are used currently. We could probably work around that, but I was trying to make the steps as generic as possible.

With the original reason for that symlink (https://github.com/moby/moby/pull/12049) no longer being in place (https://github.com/moby/moby/pull/34685), I think it should be safe to remove it, as it no longer serves any purpose.

Pushing a commit to remove it (and to see if that removes the problem), but will open it as a separate PR for visibility

thaJeztah replied:
Opened https://github.com/moby/moby/pull/39795 with just that commit

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "integration_junit" git@github.com:thaJeztah/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842359015768
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

seemethere commented on hack/make.sh:
```diff
@@ -171,13 +171,6 @@ main() {
-	if [ "$(go env GOHOSTOS)" != 'windows' ]; then
-		rm -f bundles/latest
-		# preserve latest symlink for backward compatibility
-		ln -sf . bundles/latest
```
I think we might actually use the latest symlink but let me double check

thaJeztah commented:
Will rebase again after https://github.com/moby/moby/pull/39809 is merged

thaJeztah commented:
s390x failure is a flaky test, tracked through https://github.com/moby/moby/issues/38587

https://ci.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-39724/runs/48/nodes/35/log/?start=0


```
[2019-09-03T08:41:09.641Z] === Failed
[2019-09-03T08:41:09.641Z] === FAIL: pkg/plugins TestClientWithRequestTimeout (0.00s)
[2019-09-03T08:41:09.641Z]     client_test.go:254: assertion failed: expected an error, got nil: expected error
```

thaJeztah commented:
Rebased this one, now that https://github.com/moby/moby/pull/39802 is merged; this should be ready for review/merge

tiborvass commented on Jenkinsfile:
```diff
@@ -22,6 +22,7 @@ pipeline {
         DOCKER_GRAPHDRIVER  = 'overlay2'
         APT_MIRROR          = 'cdn-fastly.deb.debian.org'
         CHECK_CONFIG_COMMIT = '78405559cfe5987174aa2cb6463b9b2c1b917255'
+        TESTDEBUG           = '1'
```
Is this temporary?

thaJeztah replied:
It gives a bit more information what's running; was in doubt to remove the commit entirely, or set it to `0`; wdyt?

thaJeztah commented:
Disabled `TESTDEBUG`, and pushed a commit to make a test fail

thaJeztah commented:
<img width="1570" alt="Screenshot 2019-09-11 at 22 46 39" src="https://user-images.githubusercontent.com/1804568/64734014-20d66980-d4e6-11e9-96a3-918f823d8a29.png">
<img width="1243" alt="Screenshot 2019-09-11 at 22 47 02" src="https://user-images.githubusercontent.com/1804568/64734015-20d66980-d4e6-11e9-902d-f1785cff8be9.png">

tiborvass commented:
@thaJeztah ok you can revert the intentional failure and i'll approve.

thaJeztah commented:
@tiborvass done 👍

thaJeztah commented:
Next steps;

- implement the same for the Windows stages
- implement the same for the integration-cli tests (now that we got rid of go-check)

thaJeztah commented:
Also (TBD); I read that the junit.xml "standard" expects test durations to be _including_ test setup/teardown. I think currently gotestsum uses the duration of the test itself (which is useful, but having timings _including_ setup/teardown is useful as well)

thaJeztah commented:
Looks like Windows RS1 nodes are running out of disk? https://ci.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-39724/runs/52/nodes/50/log/?start=0


```

[2019-09-11T21:10:43.462Z] powershell.exe : failed to register layer: re-exec error: exit status 1: output: BackupWrite 
[2019-09-11T21:10:43.462Z] \\?\D:\CI-52\CI-d723643dc\daemon\windowsfilter\9e7bef537440980fdd2c1e2a3061dc0f84770c4638b38017e34e5b2e89b09e4a\Files\Windows\INF\cpu.inf: There is not enough space on the disk.
[2019-09-11T21:10:43.462Z] At D:\gopath\src\github.com\docker\docker@tmp\durable-1485d7a6\powershellWrapper.ps1:3 char:1
[2019-09-11T21:10:43.462Z] + & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -Comm ...
[2019-09-11T21:10:43.462Z] + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2019-09-11T21:10:43.462Z]     + CategoryInfo          : NotSpecified: (failed to regis...ce on the disk.    :String) [], RemoteException
[2019-09-11T21:10:43.462Z]     + FullyQualifiedErrorId : NativeCommandError
[2019-09-11T21:10:43.462Z]  
[2019-09-11T21:10:43.462Z] 
[2019-09-11T21:10:43.462Z] 
[2019-09-11T21:10:43.462Z] ERROR: Failed 'ERROR: Failed to docker pull mcr.microsoft.com/windows/servercore:ltsc2016 into daemon under test.' at 09/11/2019 21:10:38
[2019-09-11T21:10:43.462Z] At D:\gopath\src\github.com\docker\docker\hack\ci\windows.ps1:708 char:21
[2019-09-11T21:10:43.462Z] + ...             Throw $("ERROR: Failed to docker pull $($env:WINDOWS_BASE ...
[2019-09-11T21:10:43.462Z] +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2019-09-11T21:10:43.462Z] 
```

thaJeztah commented:
^^ @StefanScherer (also wondering if that can be related to the change to use `gopath` instead of `go`; could it be that some cleanup script is not taking that path change into account?
